src/array/list/mod.rs

```diff
@@ -46,8 +46,13 @@ impl<O: Offset> ListArray<O> {
     ) -> Self {
         check_offsets(&offsets, values.len());
 
-        // validate data_type
-        let _ = Self::get_child(&data_type);
+        let nest_level = Self::get_nested_level(&data_type);
+        let expected_data_type = Self::default_datatype(values.data_type().clone());
+        let expected_nest_level = Self::get_nested_level(&expected_data_type);
```

Comment:
I think that this should be more complete:

```rust
        // validate data_type
        let data_type = Self::get_child(&data_type);
        assert_eq!(data_type, values.data_type(), "The child's datatype must match the inner type of the \'data_type\'");
```

as it avoids recursing and also proves that not only the levels are right, but that the type itself is right.

Reply:
That is a good solution. Let me change that

Reply:
I tried this approach and it fails to panic when a ListArray is nested

Reply:
That is unexpected to me: I would expect the inner `ListArray` to perform the same check when it is initialized via `from_data`.

Reply:
I thought the same, but I used your proposal to see if it will fail the nested test and it did pass without the expected panic.

Reply:
I can't reproduce your behavior; if I change to 

```
        // validate data_type
        let child_data_type = Self::get_child(&data_type);
        assert_eq!(
            child_data_type,
            values.data_type(),
            "The child's datatype must match the inner type of the \'data_type\'"
        );
```

`test_nested_panic` passes after changing the corresponding the expected to `expected = "The child's datatype must match the inner type of the \'data_type\'"`

Reply:
mmm, that is really odd. Let me see what Im doing wrong. I did just what you did and the nested test didn't fail